Adding logs to sesman/session.c

aquesnel opened the pull request:
Adding logs to sesman/session.c to help a sys admin understand why a session failed to start.

metalefty commented on sesman/session.c:
```diff
@@ -837,12 +889,25 @@ session_start_fork(tbus data, tui8 type, struct SCP_CONNECTION *c,
                 auth_end(data);
+
+                LOG(LOG_LEVEL_INFO, "Cleaning up session: terminating Xserver "
+                    "(pid %d)", display_pid);
```
What about also record the display number?

aquesnel replied:
done

metalefty commented on sesman/session.c:
```diff
@@ -837,12 +889,25 @@ session_start_fork(tbus data, tui8 type, struct SCP_CONNECTION *c,
                 g_sigterm(display_pid);
+
+                LOG(LOG_LEVEL_INFO, "Cleaning up session: terminating "
+                    "xrdp-chansrv (pid %d)", display_pid);
```
I think `display_pid` shuold be `chansrv_pid`.

aquesnel replied:
fixed

matt335672 commented on sesman/session.c:
```diff
@@ -173,6 +169,7 @@ x_server_running_check_ports(int display)
 
     if (!x_running)
     {
+        LOG(LOG_LEVEL_INFO, "Did not find a running X server at %s", text);
```
This is going to potentially generate a lot of INFO messages, as `x_server_running_check_ports()` is called in a loop, and a message is logged for each reason why the X server isn't running. You're already logging the reason why the X server isn't running at the end of the function, so I think you could remove the "Did not find..." messages and just beef up the logging at the end (i.e.):-

```c
if (x_running)
{
    if (isdigit(text[0]))
    {
        LOG(LOG_LEVEL_INFO, "Found X server active on display %d (TCP port:%s)", display, text);
    }
    else
    {
        LOG(LOG_LEVEL_INFO, "Found X server active on display %d (file:%s)", display, text);
    }
}
else
{
    LOG(LOG_LEVEL_INFO, "No X server seems to be active on display %d", display);
}
```

That can be somewhat improved on (particularly using `isdigit` to detect a port), but I hope it gives you the idea.

aquesnel replied:
The logging at the end of the function only says that last port/socket that was checked for an X server. In my past debugging I have found it very useful in the past to see all of the ports/sockets where xrdp is looking for an X server and in which order the checks are performed. So I've kept all the log messages but lowered the logging level to debug so that it's less noisy under normal circumstances.

matt335672 commented on sesman/session.c:
```diff
@@ -809,6 +848,13 @@ session_start_fork(tbus data, tui8 type, struct SCP_CONNECTION *c,
+                int chansrv_exit_code;
+
+                LOG(LOG_LEVEL_DEBUG, "session_start_fork (pid %d): waiting for "
+                    "the Xserver and the creating the channel server",
```
'then' rather than 'the'?

aquesnel replied:
done

matt335672 commented on sesman/session.c:
```diff
@@ -818,14 +864,20 @@ session_start_fork(tbus data, tui8 type, struct SCP_CONNECTION *c,
+                        "non-zero exit code (%d). Window manager config problem?",
+                        window_manager_pid, wm_exit_code);
+                }
+                else if (wm_wait_time < 10)
```
Suggest making the `wm_wait_time < 10` check unconditional, as this was explicitly introduced to catch the common case of the session exiting early. This is something I generally end up triaging, so I was hoping to be able to check logs for 'exited quickly'. Your change makes this harder to spot. How about something like:-

```c
wm_exit_code = g_waitpid_status(window_manager_pid);
wm_wait_time = g_time1() - wm_wait_time;
if (wm_exit_code != 0)
{
    LOG(LOG_LEVEL_WARNING, "window manager (pid %d) exited with "
             "non-zero exit code (%d)",
             window_manager_pid, wm_exit_code);
}
if (wm_wait_time < 10)
{
    /* Common config issue */
    LOG(LOG_LEVEL_WARNING, "window manager (pid %d) exited quickly "
             "(%d secs). Window manager config problem?",
             window_manager_pid, wm_wait_time);
}

```

aquesnel replied:
done

matt335672 commented on common/os_calls.c:
```diff
@@ -2876,6 +2876,45 @@ g_waitpid(int pid)
+    return rv;
+#endif
+}
+
```
This function doesn't cope with a process exiting after catching a signal. This is relatively common for chansrv. In such cases it looks like rv is set to zero which is misleading.

There's no great answer here I can see without changing the interface, but a better solution might be to include (after the WIFEXITED() test):-

```c
else if (WIFSIGNALED(status))
{
    /* Specify a value not representable in 8 bits */
    rv = -128 - WTERMSIG(status);
}
```

The logic is that `WEXITSTATUS()`returns an 8 bit value (either 0-255 or -128-127), so we can at least see a signal for what it is. This can probably be improved on.

aquesnel replied:
Since `g_waitpid_status()` is a new function I just changed the interface to return a struct with the return code and signal code.

matt335672 commented on common/string_calls.c:
```diff
@@ -139,6 +139,44 @@ g_text2bool(const char *s)
 
+/*****************************************************************************/
+char *
+g_strnjoin(char *dest, int dest_len, const char *joiner, const char *src[], int src_len)
```
src_len == 0 causes src[0]to be dereferenced at line 170. Probably need a guard for this (and a test?)

aquesnel replied:
fixed

matt335672 commented on common/string_calls.h:
```diff
@@ -80,6 +80,8 @@ g_bool2text(int value);
 int
 g_text2bool(const char *s);
 
+char    *g_strnjoin(char *dest, int dest_len, const char *joiner, const char *src[], int src_len);
```
Suggest Doxygen header for new funcs in this module.

aquesnel replied:
done

matt335672 reviewed (commented):
A nice illustration of check.

Couple of general points/questions:-
- I take it the plan is to use g_strjoin() in this PR? It doesn't seem to be, unless I'm missing something.
- Need to add package `check` to `.cirrus.yml`.
- I can't see why the CI is failing on Ubuntu. Any ideas??

aquesnel commented:
@matt335672 , yes the plan is to use g_strnjoin in to implement the feedback I got in the pull request. I broke up my changes into several commits and then I only submitted the first commit to check that this change is working on all platforms.

I did discover that my changes are failing because of the [-Wstringop-truncation](https://developers.redhat.com/blog/2018/05/24/detecting-string-truncation-with-gcc-8/) warning introduced in gcc 8 which is why the build only fails on the newer platforms. The warning is definitely valid and a bug that I'm going to fix. 

What I haven't figured out yet is why the makefile generated by autotools has the gcc output piped to `/dev/null` which seems to be the cause of there being no output in the build log.

aquesnel commented:
@metalefty and @matt335672 I've fixed up the code to address your feedback.

I did find out why the build was failing with no output. It turns out that the gcc `-Wstringop-truncation` check was only finding the issue when the cope was complied without `-fPIC` (position independent code). Libtool is what does the compilation of libraries, and libtool compiles the code both with and without `-fPIC` and pipes the results of compiling without `-fPIC` to `/dev/null` because the expectation is that if there is an errors from compiling the code, then there will be the same errors when compiled with and without `-fPIC`, so by default libtool hides the expected duplicate output. There is the libtool switch `-no-suppress` to change the behaviour and always show the output from both compilations.

I tried to `-no-suppress` to out config and makefiles, but as far as I can tell it can't be done. The `-no-suppress` flag is special and that it needs to be passed to libtool on the command line after the compiler path argument. The problem is that there is autotools variable which is applied to only libtool invocations which is passed after the compiler argument. The existing `AM_LIBTOOLFLAGS` and `LIBTOOLFLAGS` variables are passed to libtool before the compiler argument which makes it so that libtool fails. I also tried passing `-no-suppress` via `AM_CFLAGS` and `CFLAGS` which works for compiling the libraries but fails for  compiling the xrdp binaries (which don't use libtool).

Since this is a weird edge case and there is nothing clean that can be done, I'm just going to leave the compilation flags as is and hope that we don't keep seeing this weird edge case.

matt335672 commented:
Nice bit of detective work on the libtool issue - that's really unexpected.

I had a quick play with this. I can suggest a partial solution, which is to build just the common library with `-no-suppress`:-
```diff
--- a/common/Makefile.am
+++ b/common/Makefile.am
@@ -32,7 +32,10 @@ if XRDP_DEBUG
 AM_CPPFLAGS += -DXRDP_DEBUG
 endif
 
-AM_CFLAGS = $(OPENSSL_CFLAGS)
+# -no-suppress is an automake-specific flag which is needed
+# to prevent us missing compiler errors in some circumstances
+# (see #1843)
+AM_CFLAGS = -no-suppress $(OPENSSL_CFLAGS)
 
 module_LTLIBRARIES = \
   libcommon.la
```

I agree it's not perfect, but given most of these problems are likely to be located in `string_calls.c` it might be worth considering.

Review to follow...

matt335672 commented on sesman/session.c:
```diff
@@ -396,9 +400,8 @@ session_start_chansrv(char *username, int display)
         /* should not get here */
-        LOG(LOG_LEVEL_ERROR, "error starting chansrv "
-            "- user %s - pid %d", username, g_getpid());
         list_delete(chansrv_params);
```
Don't understand why the `LOG()` is removed. We should have a log if the `g_evecvp()` fails shouldn't we?

aquesnel replied:
I moved all of the error logging about failed `exec` calls to the `g_exec*` functions. See [os_calls.c](https://github.com/neutrinolabs/xrdp/pull/1843/commits/b530b9fa5b4d56b12d89fade9c3e2ed25d7d0e86#diff-b04475dad2718b002e94b443cb48a735bae669abe6ab10bdfc10f17f7adf1a20R2616-R2627)

matt335672 replied:
That's a much better idea.

matt335672 reviewed (commented):
Thanks for the updates to this. Let me know if I haven't been clear on any of the above comments.

metalefty commented on .cirrus.yml:
```diff
@@ -6,7 +6,7 @@ FreeBSD_task:
       image_family: freebsd-12-2
   prepare_script:
-    - pkg install -y $SSL git autoconf automake libtool pkgconf opus jpeg-turbo fdk-aac pixman libX11 libXfixes libXrandr nasm
+    - pkg install -y $SSL git autoconf automake libtool pkgconf opus jpeg-turbo fdk-aac pixman libX11 libXfixes libXrandr nasm check
```
Thanks for the catch!

aquesnel commented on common/string_calls.c:
```diff
@@ -651,3 +651,42 @@ g_strtrim(char *str, int trim_flags)
+
+    if (dest == NULL || dest_len < 1 || src == NULL || src_len < 1)
+    {
+        return dest;
```
done

aquesnel commented on common/string_calls.c:
```diff
@@ -651,3 +651,42 @@ g_strtrim(char *str, int trim_flags)
+    {
+        len = g_strlen(src[i]);
+        g_strncat(dest_pos, src[i], dest_end - dest_pos);
+        dest_pos += len;
```
Thanks, I had no idea about this limitation.

aquesnel commented on common/os_calls.c:
```diff
@@ -2901,17 +2953,21 @@ g_waitpid_status(int pid)
-    return rv;
+    return exit_status;
 #endif
 }
```
fixed

aquesnel commented on common/os_calls.h:
```diff
@@ -23,6 +23,12 @@
 
 #include "arch.h"
 
+struct exit_status
```
good idea, fixed.

aquesnel commented on scripts/install_xrdp_build_dependencies_with_apt.sh:
```diff
@@ -1,20 +1,18 @@
-
-if [ -z "$ARCH" ]; then
-    ARCH=amd64
+if [ $# -ge 2 ]; then
```
you are correct, fixed now.

aquesnel commented on sesman/session.c:
```diff
@@ -568,96 +573,84 @@ session_start_fork(tbus data, tui8 type, struct SCP_CONNECTION *c,
+                    }
+                    else
+                    {
+                        LOG(LOG_LEVEL_DEBUG, 
```
Here is my thought process:
* the sys admin controls the xrdp config file to enable/disable the `EnableUserWindowManager` setting
* the user controls the `startwm.sh` window manager file in their home directory, including if it exists or not

So I'd expect it to be a normal state for the `startwm.sh` file to be missing from the home directory.

matt335672 replied:
Agreed - good explanation.

aquesnel commented on sesman/session.c:
```diff
@@ -816,98 +808,105 @@ session_start_fork(tbus data, tui8 type, struct SCP_CONNECTION *c,
+                wm_exit_status = g_waitpid_status(window_manager_pid);
                 wm_wait_time = g_time1() - wm_wait_time;
-                if (wm_exit_code != 0)
+                if (wm_exit_status.exit_code != 0)
```
fixed

aquesnel commented on sesman/session.c:
```diff
@@ -816,98 +808,105 @@ session_start_fork(tbus data, tui8 type, struct SCP_CONNECTION *c,
+                        "was running for %d seconds.",
+                        window_manager_pid, display, wm_wait_time);
                 }
-                LOG(LOG_LEVEL_INFO, "Cleaning up session. Calling "
```
good point. fixed

aquesnel commented on sesman/session.c:
```diff
@@ -482,12 +484,16 @@ session_start_fork(tbus data, tui8 type, struct SCP_CONNECTION *c,
     {
-        LOG(LOG_LEVEL_INFO, "calling auth_start_session from pid %d",
-            g_getpid());
+        LOG_DEVEL(LOG_LEVEL_DEBUG, 
```
good point. fixed

aquesnel commented:
Hi @matt335672 ,

I've fixed up the code to address the feedback. There are a few of your comments that I responded to, so please let me know if you want me to make additional changes.

I've also rebased and squashed the changes onto devel.

matt335672 commented on common/string_calls.c:
```diff
@@ -775,3 +776,50 @@ g_strtrim(char *str, int trim_flags)
+        g_strncat(dest_pos, src[i], dest_end - dest_pos);
+    }
+
+    dest[dest_len] = '\0';
```
Off by one - Should be `dest[dest_len - 1] = '\0`, or `*dest_end = '\0';`

This test catches it:-

```c
START_TEST(test_strnjoin__doesnt_write_off_end)
{
    /* setup */

    const char *src[] = { "a","b","c"};
    char result[5+1+1]; /* a-b-c + term + guard value */

    /* test */

    result[6] = '\x7f';
    g_strnjoin(result, 5 + 1, "-", src, 3);

    /* verify */

    ck_assert_int_eq(result[6], '\x7f');
}
END_TEST
```
Alternatively remove the whole line. You're using `strncat()` so this terminator is guaranteed to be present anyway.

BTW - nice use of MIN().

aquesnel replied:
you are right on both points. I added the test case you wrote, and removed the unnecessary writing null after the end of the destination buffer.

I had added the null termination because I misread `strncpy` for `strncat` in the following article:

>An entirely different form of string truncation is one that can result from calls to strncpy. Unlike functions such as snprintf and strncat that always append a terminating NUL, when the source string passed to strncpy is longer than the size specified by the third argument, the function truncates the copy without appending a NUL to the end. 
From [Forming Non-NUL-Terminated Sequences](https://developers.redhat.com/blog/2018/05/24/detecting-string-truncation-with-gcc-8/)

matt335672 reviewed (changes requested):
Good job. Please check my reasoning on the comment in g_strnjoin(). Also you'll need to rebase for which I apologise.

aquesnel commented:
@matt335672 I think this pull request is ready to merge now. Please let me know if there is anything else that needs to be fixed up.

matt335672 reviewed (approved):
Thanks for the swift turnaround.

matt335672 commented:
@aquesnel - I managed to introduce a conflict by merging #1858 before this one! It seemed a bit rotten of me to ask you to rebase yet again, so I've added a merge commit to fix things.

Thanks so much for adding the check stuff - I think that's going to be very useful in the future.

metalefty commented on configure.ac:
```diff
@@ -357,6 +357,11 @@ AC_CHECK_HEADER([X11/extensions/Xrandr.h], [],
+# checking for libcheck (used for unit testing)
+PKG_CHECK_MODULES([CHECK], [check >= 0.10.0])
+AC_CHECK_HEADER([check.h], [],
+  [AC_MSG_ERROR([please install check])])
```
I noticed that this makes `check` is always required to build xrdp even when users try to install xrdp to their system. I don't think unit testing is always required. Performing unit tests during CI is enough. What do you think?

matt335672 replied:
Yes, I think that's a great idea. I see you've raised #1860 - I'll take a look.
